Fix/Replace Rect With Square

vancouver29 opened the pull request:
This method returns a Size which is always positive and even

vancouver29 committed:
updated new(exp) method of Size

This method returns a Size which is always positive and even

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -185,9 +186,23 @@ pub struct Size {
+
+        //if exp < 0 { panic!("An Exponent must be positive! "); }
+        let width = u32::pow(2, exp);
+        let height = u32::pow(2, exp);
```
Commented out code here. Also, since we already made the calculation `u32::pow(2, exp)` once, you can copy the value into height:

```
let width = u32::pow(2, exp);
let height = width;
```

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -223,13 +238,23 @@ pub struct Rect {
+//            origin: Point::new(x, y),
+//            size: Size::new(ex),
+//        }
+//    }
```
Commented out code here.

vancouver29 replied:
what do u mean ? I should delete these lines of code ?

johnxnguyen replied:
Are you going to use this code?

vancouver29 replied:
im not going to use this code , but maybe you need them later ?

johnxnguyen replied:
You can delete it.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -254,21 +279,32 @@ impl Rect {
+//            let half_width = n.width >> 1;
+//            let half_height = n.height >> 1;
+//            (half_height,half_width)
+//        };
```
Commented out code here.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
-    }
+//    #[test]
+//    #[should_panic(expected = "A size's width and/or height must be positive.")]
+//    fn size_non_positive_width() {
```
Commented out code here.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
-    }
+//    #[test]
+//    #[should_panic(expected = "A size's width and/or height must be positive.")]
+//    fn size_non_positive_height() {
```
Commented out code here.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
-    }
+//    #[test]
+//    #[should_panic(expected = "A size's width and/or height must be positive.")]
+//    fn rect_non_positive_size() {
```
Commented out code here.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
+        assert_eq!(NW(Rect::new(4.0, 4.0, Size::new_wh(2,2))), nw);
+        assert_eq!(NE(Rect::new(6.0, 4.0, Size::new_wh(2,2))), ne);
+        assert_eq!(SW(Rect::new(4.0, 2.0, Size::new_wh(2,2))), sw);
+        assert_eq!(SE(Rect::new(6.0, 2.0, Size::new_wh(2,2))), se);
```
Can't you use `Size::new(1)` to create a `Size` with width and height 2?

vancouver29 replied:
yeah you are right , i just forgot to change them, because I used those lines to test

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -185,9 +186,23 @@ pub struct Size {
+    pub fn new_wh(width: u32, height: u32) -> Size {
+        Size{width, height}
     }
 }
```
Is this method necessary? It makes it possible to create sizes with unequal width and height, as well as width and heights that aren't a power of 2, which is something we want to restrict.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
 
     #[test]
     fn rect_even_quadrants() {
         // given
```
I'd suggest to rename this test to `rect_quadrants()` since that is all it is testing. The name `rect_even_quadrants()` refers to when the width and height could be even / odd.

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -457,76 +493,76 @@ mod tests {
     fn rect_uneven_quadrants() {
         // given
-        let sut = Rect::new(-5.0, -5.0, 5, 5);
+        let sut = Rect::new(-5.0, -5.0, Size::new(5));
```
This test is redundant, see previous comment.

johnxnguyen reviewed (changes requested):
It would be nice to have a test for `Size::new(exp:)`.

vancouver29 committed:
redundant codes

- redundant codes are removed
- rect_quadrants() replaced rect_even_quadrants()
- test case are updated

vancouver29 committed:
height gets the value of width

johnxnguyen commented on src/geometry/types.rs:
```diff
@@ -185,10 +186,21 @@ pub struct Size {
+    pub fn new(exp: u32) -> Size {
+
+        //if exp < 0 { panic!("An Exponent must be positive! "); }
+        let width = u32::pow(2, exp);
```
There's still commented out code here.

johnxnguyen commented:
Don't forget to test `Size::new(exp:)`. Also, you're build is failing because you deleted the `Size::new_wh(width:height:)` method but you are still using it in your code.

vancouver29 committed:
update values for Points in Test

vancouver29 committed:
rect_quadrants_of_unit_rect()

johnxnguyen committed:
Fix remaining tests

johnxnguyen committed:
Tidy up

johnxnguyen committed:
Remove Size

johnxnguyen committed:
Merge branch 'develop' into feature/edited_Size

johnxnguyen committed:
Rename Rect to Square

johnxnguyen committed:
Merge branch 'feature/edited_Size' of github.com:johnxnguyen/Newton into feature/edited_Size
